### DISCUSSION TITLE
Custom required action provider not getting trigger

### DISCUSSION THREAD

**Discussion by Bhavik237** [keycloak/keycloak#21145] [Archive]
Hello , everyone

After developing a custom Required Action, which triggers a change in the user's attribute, I have successfully added the custom required action in Keycloak and associated it with the user. However, upon the user's login, the custom action does not seem to be triggered. What could be the possible reasons for this behavior?
keycloak version 21.1.1 using
code 
`
public class CustomReqAction implements RequiredActionProvider {
    public static final String PROVIDER_ID = "custom-action";

    @Override
    public void evaluateTriggers(RequiredActionContext context) {

    }

    @Override
    public void requiredActionChallenge(RequiredActionContext context) {
        
    }

    @Override
    public void processAction(RequiredActionContext context) {
        logger.debug("changing user attribute -----------");
        UserModel user = context.getUser();
        user.setSingleAttribute("tobereset", "no");
        logger.debug(user.getAttributes());
        context.success();
    }

    @Override
    public void close() {
        // Not used in this example
    }
}
`
Provider Factory
`public class CustomReqActionFactory implements RequiredActionFactory {

    @Override
    public String getDisplayText() {
        return "Custom Required action";
    }

    @Override
    public RequiredActionProvider create(KeycloakSession session) {
        return new CustomReqAction();
    }

    @Override
    public void init(Config.Scope config) {

    }

    @Override
    public void postInit(KeycloakSessionFactory factory) {

    }

    @Override
    public void close() {

    }

    @Override
    public String getId() {
        return CustomReqAction.PROVIDER_ID;
    }
}
`

Please assist me.
Thanks

**Comment by dasniko** [comment]
Keycloak first calls the `requiredActionChallenge()` method, the `processAction()` method is only called after the user submitted the form, which was possibly set as a challenge in the former method.
So, if you don't have a user interaction (IMHO thus you shouldn't use required action, as this is meant to have user interaction), you should put your code into the `requiredActionChallenge()` method.

**Reply by Bhavik237** [reply]
Thanks , now it is working as expected.
